FIX: Replace R2 gem with rtlcss for generating RTL CSS

OsamaSayegh opened the pull request:
We've had a couple of problems with the R2 gem where it generated a broken RTL CSS bundle that caused a badly broken layout when Discourse is used in an RTL language, see #18446 and #19602. For this reason, we're replacing R2 with `rtlcss` that can handle modern CSS features better than R2 does.

Internal topic: t/76263.

SamSaffron commented on app/assets/javascripts/package.json:
```diff
@@ -14,6 +14,7 @@
     "truth-helpers",
-    "wizard"
+    "wizard",
+    "rtlcss-miniracer"
```
are we sure we want it in our app/assets ? will it be included in our bundle somehow?

OsamaSayegh replied:
I don't think it would sneak into app JS bundle for production, but there's no reason to put this in our main `package.json` so I've removed it.

SamSaffron commented:
I am approving, this is fantastic work. 

But we need to confirm this is not somehow sneaking into our application client bundle. cc @davidtaylorhq

CvX reviewed (commented):
Other than the inline notes, I don't really like adding build artifacts (`dist/main.js`) to the repo but I'm not sure where would be the good place for an automated `yarn build`/`yarn webpack` step 🤔 @davidtaylorhq any ideas?

OsamaSayegh commented on lib/stylesheet/compiler.rb:
```diff
@@ -65,14 +65,27 @@ def self.compile(stylesheet, filename, options = {})
+        end
+      @context.eval("rtlcss.default.process(#{src_css.inspect})")
+    rescue MiniRacer::RuntimeError
+      nil
```
We RTLify only core CSS and we do it during builds/deploys so I guess we do want to know about failures if they ever happen, though I don't think will. I've removed the rescue for now, but we can add it later if turns out we actually need it.

davidtaylorhq commented:
> I don't really like adding build artifacts (dist/main.js) to the repo

Yeah agreed! We could potentially slot in a build step somewhere, but it's gonna slow down builds (and development).

How crazy would it be to create an `r2_wrapper` gem which implements `Stylesheets::Manager.rtlify_css`. All the webpack stuff could be included in that gem's repository. Core would then simply install the gem and call `R2Wrapper.rtlify_css()`.

The build artefact could either be included in the gem's repository, or could be generated just before pushing to Rubygems.

davidtaylorhq commented on lib/stylesheet/compiler.rb:
```diff
@@ -65,14 +65,25 @@ def self.compile(stylesheet, filename, options = {})
+    def self.rtlify_css(src_css)
+      @context ||=
+        begin
+          context = MiniRacer::Context.new
```
I think we should add a timeout here for safety. It can be quite high - just need a safety net for any pathalogical bugs. Would 10 seconds be reasonable?

davidtaylorhq commented on lib/stylesheet/compiler.rb:
```diff
@@ -65,14 +65,25 @@ def self.compile(stylesheet, filename, options = {})
+          )
+          context
+        end
+      @context.eval("rtlcss.process(#{src_css.inspect})")
```
`.inspect` does not always make an object safe to interpolate into JS. `.to_json` does.
```suggestion
      @context.eval("rtlcss.process(#{src_css.to_json})")
```

davidtaylorhq replied:
Or even better - use `@context.call` to avoid the need for interpolation and improve performance: https://github.com/rubyjs/mini_racer#function-call

(In core, we do this a lot in [discourse_js_processor.rb](https://github.com/discourse/discourse/blob/b96869d5fb7702f74b8191d3fb0108d68ff0df51/lib/discourse_js_processor.rb#L252-L252))

OsamaSayegh commented:
Thank you for the great feedback @davidtaylorhq! I've moved everything to https://github.com/discourse/rtlcss_wrapper (including the changes you suggested here) and added `rtlcss_wrapper` to the `Gemfile` here. Let me know what you think. cc @SamSaffron @CvX since you've commented on this PR.
